Add Optimize Images setting and enable it by default

fluiddot opened the pull request:
## Description

This PR adds the Optimize Images setting to the App Settings screen, similar to what we have in the Android version.

**Media Settings added/updated:**
* **Optimize Images:** When turning on, max image upload size and image quality will be applied to images upon upload. When it's off, images will be uploaded with high quality and their original size.
* **Max Image Upload Size:** Maximum size allowed in image uploades. If the image to upload is bigger, it will be scaled to the value set as maximum. **_This setting won't be shown if Optimize Images setting is off._**
* **Image Quality:** Quality to be applied to image uploads. **_This setting won't be shown if Optimize Images setting is off._**

**By default, the Optimize Image setting will be on** and we'll apply the following default values to other media settings:
* Max Image Upload Size: 2000px
* Image Quality: Medium

**NOTE:** If the user already previously set a value in _Max Image Upload Size_ setting, it will remain that value.

Additionally, in the following upload flows, the app will ask the user to keep or disable the Optimize Images setting. The popup will only be shown once.
* ~~Media screen - Uploading using _Choose from My Device/Take a Photo_ option.~~ The media screen is now using `SiteMediaViewController` (https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#issuecomment-1821981326).
* New Media Screen (`SiteMedia`) - Uploading using _Choose from My Device/Take a Photo_ option.
**NOTE:** This new screen is only available via the _Media Modernization_ feature flag.
* Gutenberg Editor - When using specific blocks and using _Choose from My Device/Take a Photo_ option.
  * Image block
  * Gallery block
  * Cover block
  * Media & Text block
* Gutenberg Editor - Adding/updating the featured image and using _Choose from My Device/Take a Photo_ option.

### Screenshots

| Default values | Optimize Images popup |
|--------|--------|
| <img width=300 src=https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/24aa0193-8017-4b5a-bd9d-13fd9d90d89b>  | <img width=300 src=https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/7ff64db6-e1bc-44ca-b93e-5ea876213a26> | 

### Video captures

| Media settings | Optimize Images popup |
|--------|--------|
| <video src=https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/b1d00db1-4a11-4db1-bfa1-fa8470ca8373> | <video src=https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/6a7c1869-6cfd-4b14-94a8-707b0717b0e3> | 

## To test
### Default setting should be changed if it has never been set before

* Perform a fresh installation of the app.
* Navigate to _Me → App Settings_. 
* Observe that _Optimize Images_ setting is enabled by default.
* Observe that _Max Image Upload Size_ and _Image Quality_ settings have the default values as shared in the PR's description.

### Default setting should not be changed if it has been set previously

#### Set original as max image upload size
* Install an older version.
* Navigate to _Me → App Settings_. 
* Move the _Max Image Upload Size_ setting to a different value than Original.
* Move the _Max Image Upload Size_ setting back to the Original value.
* Install the installable build from this PR.
* Navigate to _Me → App Settings_. 
* Observe that the _Optimize Images_ setting is enabled and that _Max Image Upload Size_ setting remains in the previously set Original value.

#### Set max image upload size
* Install an older version.
* Navigate to _Me → App Settings_. 
* Move the _Max Image Upload Size_ setting to a different value than Original.
* Install the installable build from this PR.
* Navigate to _Me → App Settings_. 
* Observe that the _Optimize Images_ setting is enabled and that _Max Image Upload Size_ setting remains in the previously set value.

### Resize settings should function as expected
- Add an image to the device with dimensions above 2000px.
- Remove and install the app.
- Navigate to the Posts/Pages screen.
- Open/create a post/page.
- Add an Image block.
- Tap on the _Choose from device_ option.
- Select the image previously added with large dimensions.
- When the _Optimize Images_ pop-up is displayed, select the option to keep image optimization.
- Wait for the image upload to finish.
- Go to the browser and navigate to the Media screen of the site.
- Select the uploaded image and click on the _Edit_ button.
- Observe that dimensions displayed under the _DIMENSIONS_ section are lower than the original ones. The dimensions shouldn’t exceed the value set in _Max Image Upload Size_ setting.
- Observe that the file size under the _FILE SIZE_ section is smaller than the original file size.

### Compression settings should function as expected
- Add an image to the device with a large file size.
- Remove and install the app.
- Navigate to the Posts/Pages screen.
- Open/create a post/page.
- Add an Image block.
- Tap on the _Choose from device_ option.
- Select the image previously added with a large file size.
- When the _Optimize Images_ pop-up is displayed, select the option to keep image optimization.
- Wait for the image upload to finish.
- Go to the browser and navigate to the Media screen of the site.
- Select the uploaded image and click on the _Edit_ button.
- Observe that the file size under the _FILE SIZE_ section is smaller than the original file size.

### Pop-up should be displayed in specific sources

#### Media Screen
- Remove and install the app.
- Navigate to `My Site → Media`.
- Tap on ➕ button and select `Choose from My Device` option.
- Select one or multiple items.
- Tap on the `Add` button.
- Observe that the Optimize Images pop-up is displayed.

**NOTE:** The popup will be displayed independently to the media type. I.e. it could be shown when only uploading videos.

#### Gutenberg Editor

**Preparation:**
- Remove and install the app.
- Navigate to `My Site → Posts` (or Pages).
- Create a post.

##### Image Block - From Device
- Add any of the following blocks:
  - Image block
  - Gallery block
  - Cover block
  - Media & Text block
- Tap on the `Choose from device` option.
- Select one item (or multiple items when using Gallery block).
- Observe that the Optimize Images pop-up is displayed.

##### Image Block - Take Photo
- Add any of the following blocks:
  - Image block
  - Gallery block
  - Cover block
  - Media & Text block
- Tap on the `Take a Photo` option.
- Take a photo and use it.
- Observe that the Optimize Images pop-up is displayed.

#### Featured Image - From Device
- Tap on the `...` button located on the top bar.
- Tap on the `Post Settings` options.
- Navigate to the Featured Image section and tap on `Set Featured Image`.
- Tap on the `Choose from Device` option.
- Select one time.
- Observe that the Optimize Images pop-up is displayed.

#### Featured Image - Take Photo
- Tap on the `...` button located on the top bar.
- Tap on the `Post Settings` options.
- Navigate to the Featured Image section and tap on `Set Featured Image`.
- Tap on the `Take a Photo` option.
- Select one time.
- Observe that the Optimize Images pop-up is displayed.

### Check events

Observe that the following events are sent:

- `APP_SETTINGS_OPTIMIZE_IMAGES_CHANGED`
  - Source: Changing _Optimize Images_ setting in _App Settings_ screen.
  - Property `enabled`: `True` if the app is optimizing images for uploads.
- `APP_SETTINGS_MAX_IMAGE_SIZE_CHANGED`
  - Source: Changing _Max Image Upload Size_ setting in _App Settings_ screen.
  - Property `size`: Max size used.
- `APP_SETTINGS_IMAGE_QUALITY_CHANGED`
  - Source: Changing _Image Quality_ setting in _App Settings_ screen.
  - Property `quality`: Quality value used.
- `APP_SETTINGS_OPTIMIZE_IMAGES_POPUP_TAPPED`
  - Source: When the Optimize Images popup is displayed, tap on one of the two options. The popup can be triggered by removing and reinstalling the app and then uploading an image.
  - Property `option`: `on` if the user decided to leave image optimization on. `off` if the user wants to disable it.

## Regression Notes
1. Potential unintended areas of impact
These changes mainly affect the App Settings screen and media upload flows.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
* Manual testing.
* [`MediaSettingsTests`](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/WordPressTest/MediaSettingsTests.swift)

3. What automated tests I added (or what prevented me from doing so)
* Expanded [`MediaSettingsTests`](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/WordPressTest/MediaSettingsTests.swift) to cover the new changes in `MediaSettings` file.
* I plan in a separate PR to incorporate UI tests to cover the _Optimize Images_ popup.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

fluiddot committed:
Add optimize image popup

fluiddot committed:
Add image quality setting to App settings screen

fluiddot committed:
Add image optimization setting to App Settings creen

fluiddot committed:
Disable max image setting when image optimization is off

fluiddot committed:
Set optimized value as default for max image size

fluiddot committed:
Use image quality setting when compressing images before upload

fluiddot committed:
Add new event for tracking max image size setting changes

fluiddot committed:
Display optimize images popup when uploading from device media

fluiddot committed:
Display optimize images popup when uploading after taking a photo

fluiddot committed:
Fix typos in `MediaCoordinator`

fluiddot committed:
Add inline comment to image size for upload calculation

fluiddot committed:
Move media settings default values to constants

fluiddot committed:
Update media setting unit test

fluiddot committed:
Remove unneeded self reference in `advertiseImageOptimization` handlers

kean commented on WordPress/Classes/Services/MediaSettings.swift:
```diff
@@ -3,14 +3,77 @@ import AVFoundation
+            case .medium:
+                return NSLocalizedString("Medium", comment: "Indicates an image will use medium quality when uploaded.")
+            case(.low):
+                return NSLocalizedString("Low", comment: "Indicates an image will use low quality when uploaded.")
```
80%, which is the current default, is [considered](https://sirv.com/help/articles/jpeg-quality-comparison/) optimal for web uploads. I would suggest going with 10% increments where 80% would be "Default", "Standard" or something like that.

fluiddot replied:
Thanks @kean for sharing this useful info 🙇 ! To be honest, I merely chose these quality values based on the ones we have in the Android version ([reference](https://github.com/wordpress-mobile/WordPress-Android/blob/41fd7a47cff9475ea1615da960c700b26a1033e6/WordPress/src/main/res/values/key_strings.xml#L366-L372)). However, as you suggested, it might be interesting to use different values and more close to standards:
- Maximum/Very High: 100%.
- High: 90% (following [this article](https://sirv.com/help/articles/jpeg-quality-comparison/), this JPEG quality gives a very high-quality image while gaining a significant reduction on the original 100% file size).
- Medium (default): 80% (following [this article](https://sirv.com/help/articles/jpeg-quality-comparison/), this JPEG quality gives a greater file size reduction with almost no loss in quality).
- Low: 70% (following [this article](https://sirv.com/help/articles/jpeg-quality-comparison/), 75% JPEG quality and lower begins to show obvious differences in the image, which can reduce your website user experience).

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/b6f01271e0e5eda8b0c7d610bcbe270583883dfb.

fluiddot replied:
Note that I removed `Very High` option as the above four options should be enough to set different quality settings.

fluiddot commented:
@twstokes I'm assigning you as a reviewer in case you could take a preliminary review of the PR. I'm mostly looking for feedback regarding the approach, particularly for https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/files#diff-45d2b2ed5b4684d81e4933cefdc647b6090ac127a7dd7e5ff11f5bdee894aa12R265-R280 related to animating table view cells.

The PR is almost ready, I'm just waiting to incorporate some unit tests and ~~register the events in Tracks~~ (events can be registered later on).

Thanks for the help 🙇 !

fluiddot committed:
Add `MediaSettings` unit tests

fluiddot committed:
Update release notes

fluiddot committed:
Update image quality values

The Very High option has been removed and the rest of values have been updated following more standard industry values (https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#discussion_r1383651268).

fluiddot committed:
Merge branch 'trunk' into add/image-optimization-setting

twstokes commented on WordPress/Classes/Services/MediaHelper.swift:
```diff
@@ -67,6 +67,37 @@ class MediaHelper: NSObject {
+            return
+        }
+
+        let title = NSLocalizedString("Keep optimizing images?",
```
Could we use the reverse-DNS naming style for localized strings? ([localization docs](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/docs/localization.md#localization))

fluiddot replied:
Oh, I didn't know about this, thanks for sharing it 🙇. I'll update the added/updated strings to use this style.

fluiddot replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/b968b02e854f6674a1d503530e77e7e5112cfa63.

twstokes commented:
👋 Hey @fluiddot! I've started a review and should be able to complete it on Nov 8th. Thanks!

provenself commented:
Editorial (copy review) checking in! Just offering a few suggestions for the **Optimize Images popup** modal:

> **Keep optimizing images?**
> Image optimization shrinks images for faster uploading.
> This option is enabled by default, but you can
> change it in the app settings at any time.
> Yes, leave on
> No, turn off

(I'd also suggest reversing the button order as above, so that "Yes, leave on" appears first. By requiring people to think a bit more to opt out, it subtly suggests optimization as the stronger option.)

SiobhyB reviewed (approved):
I want to spend a bit more time testing the functionality of the compression and resize settings, but wanted to share the results of the manual tests I've carried out so far relating mainly to the new pop-up.

I verified that the default is _on_ for image optimisation, that the pop-up displays in most areas, and that the track events are firing. The only media block where I don't see the pop-up so far is the Story block. I don't think that needs to be a blocker to this PR (the Android equivalent was merged with the known issue that the pop-up doesn't display when taking a photo from the Story block). But, it might be nice to track this as a lower priority task to work on separately.

### General functionality

| Test  | Result |
| ------------- | ------------- |
| Default setting should be changed if it has never been set before | ✅ |
| Pop-up _doesn't_ display when settings have been changed | ✅ |
| Pop-up _doesn't_ display after it's been dismissed | ✅ |

### Pop-up should be displayed in specific sources

| Test  | Result |
| ------------- | ------------- |
| Media Screen - From Device | ✅ |
| Media Screen - Take Photo | ✅ |
| New Media Screen | ✅ |
| Image Block - From Device | ✅ |
| Image Block - Take Photo | ✅ |
| Gallery Block - From Device | ✅ |
| Gallery Block - Take Photo | ✅ |
| Cover Block - From Device | ✅ |
| Cover Block - Take Photo | ✅ |
| Media & Text Block - From Device | ✅ |
| Media & Text Block - Take Photo | ✅ |
| Story Block - From Device | ⚠️ |
| Story Block - Take Photo | ⚠️ |
| Featured Image - From Device | ✅ |
| Featured Image - Take Photo | ✅ |

### Check events

| Test  | Result |
| ------------- | ------------- |
| `APP_SETTINGS_OPTIMIZE_IMAGES_CHANGED` | ✅ |
| `APP_SETTINGS_MAX_IMAGE_SIZE_CHANGED` | ✅ |
| `APP_SETTINGS_IMAGE_QUALITY_CHANGED` | ✅ |
| `APP_SETTINGS_OPTIMIZE_IMAGES_POPUP_TAPPED` | ✅ |

kean commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -395,11 +455,23 @@ private extension AppSettingsViewController {
     func mediaTableSection() -> ImmuTableSection {
         let mediaHeader = NSLocalizedString("Media", comment: "Title label for the media settings section in the app settings")
 
+        let imageOptimizationValue = Bool(MediaSettings().imageOptimizationSetting)
```
Please, use a multiplication sign (`×`) to display dimensions.

kean replied:
I opened a PR https://github.com/wordpress-mobile/WordPress-iOS/pull/22019. I also added thin spaces.

kean committed:
Use multiplication sign to display dimenstions in upload settings

SiobhyB reviewed (commented):
Following on from my last comment, I've further tested the functionality of the new settings and confirmed all works as expected for me. Here are some of the scenarios I verified:

* Selecting the "no, turn off" option in the pop-up disables optimization, so that the image is uploaded at its full quality and size.
* Selecting the "yes, keep on" option in the pop-up keeps the new default in place, so that the image is upload at "medium" quality and a max of 2,000px.
* Changing the defaults via App Settings work as expected in regards to how they change the size and quality of uploaded images.

When the copy is updated based on [Editorial's suggestion](https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#issuecomment-1802410807), the changes will look good to me from a functionality stand point. The code also looks good to my eyes, but I'd lean towards Alex or Tanner for their iOS expertise and advocate for a second review.

twstokes commented on WordPress/Classes/Services/MediaSettings.swift:
```diff
@@ -3,14 +3,70 @@ import AVFoundation
+    fileprivate let advertiseImageOptimizationKey = "SavedAdvertiseImageOptimization"
 
+    fileprivate let defaultImageOptimization = true
+    fileprivate let defaultMaxImageSize = 2000
```
Nit: Maybe we use `defaultMaxImageDimension` just to keep some convention with `minImageDimension` and `maxImageDimension`?

I also like using "dimension" over "size" so the reader easily knows we're talking about resolution rather than file size.

kean replied:
My programmer brain wants to replace this with 2048 🙈

fluiddot replied:
> I also like using "dimension" over "size" so the reader easily knows we're talking about resolution rather than file size.

@twstokes This makes me wonder if we should also update the label of this setting. It currently says `Max Image Upload Size`, which as you pointed out, might be interpreted as the file size, not dimensions.

<img width="250" alt="Screenshot 2023-11-10 at 19 46 32" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/ca3e632c-e77d-47f2-81ac-68718d0b67f3">

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/477777cbb71f90fe99b205de6a6d35543343705a.

twstokes commented on WordPress/Classes/Services/MediaSettings.swift:
```diff
@@ -160,12 +222,56 @@ class MediaSettings: NSObject {
         }
     }
+
+    var imageOptimizationSetting: Bool {
```
Optional naming nit: `imageOptimizationEnabled`

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/92d4d1308872176acbf5148f75e1d007bb2515dd.

twstokes commented:
Thanks for the suggestions @provenself!

> (I'd also suggest reversing the button order as above, so that "Yes, leave on" appears first. By requiring people to think a bit more to opt out, it subtly suggests optimization as the stronger option.)

Regarding this one, I'd lean towards following the [Apple Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/alerts) so we're consistent with the platform:

> In general, place the button people are most likely to choose on the trailing side in a row of buttons or at the top in a stack of buttons.

We can set "Yes, leave on" as the [`preferredAction`](https://developer.apple.com/documentation/uikit/uialertcontroller/1620102-preferredaction) of the `UIAlertController` to emphasize it with a bold font:

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/2092798/3432d4ed-395f-48b8-867d-739499cb9d70" width="300" />

twstokes commented on WordPress/Classes/Services/MediaHelper.swift:
```diff
@@ -67,6 +67,37 @@ class MediaHelper: NSObject {
+        let turnOffTitle = NSLocalizedString("No, turn off", comment: "Title of button for turning off image optimization, displayed in the alert informing users to enable image optimization in uploads.")
+        let leaveOnTitle = NSLocalizedString("Yes, leave on", comment: "Title of button for leaving on image optimization, displayed in the alert informing users to enable image optimization in uploads.")
+
+        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
```
As noted in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#issuecomment-1804985534, let's consider making "Yes, leave on" as the preferred action to get some UI / UX emphasis.

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/d64232cc9846358f242956177f7829ac23b2460d.

twstokes replied:
Looks great!

twstokes commented on WordPress/Classes/Services/MediaSettings.swift:
```diff
@@ -3,14 +3,70 @@ import AVFoundation
+            }
+        }
+
+        static func imageQuality(from value: Float) -> MediaSettings.ImageQuality {
+            switch value {
+            case 1.0:
+                return .maximum
+            case 0.9:
+                return .high
+            case 0.8:
+                return .medium
+            case 0.7:
+                return .low
+            default:
+                return .medium
+            }
+        }
```
I couldn't find where we used this anywhere. Do we need it?

fluiddot replied:
You're right, it's not currently used. I created this function following the structure we have for the video resolution. However, since we won't use it for now we could remove it.

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/a85f4b11e14911ba81b0535d8e19f74ca25f8795.

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -220,6 +253,33 @@ class AppSettingsViewController: UITableViewController {
+            MediaSettings().imageOptimizationSetting = value
+
+            // Track setting changes
+            WPAnalytics.track(.appSettingsOptimizeImagesChanged, properties: ["enabled": value as
+                                                                                     AnyObject])
```
Can we drop the cast here?

```suggestion
            WPAnalytics.track(.appSettingsOptimizeImagesChanged, properties: ["enabled": value])
```

fluiddot replied:
Sure! I used the casting here following other calls to analytics we have along the file ([example](https://github.com/wordpress-mobile/WordPress-iOS/blob/a51fb8fea237faf4b0691484fc3d84d99a787c62/WordPress/Classes/ViewRelated/Me/App%20Settings/AppSettingsViewController.swift#L252)). But I agree that seems unnecessary based on the `properties` type. I'll remove the casting from others as well 👍.

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/f29f4d2837d4a1d17b2d4f182f178680d8df74b6.

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -206,6 +205,40 @@ class AppSettingsViewController: UITableViewController {
+                // Track setting changes
+                var properties = [String: AnyObject]()
+                properties["quality"] = newQuality.description as String as AnyObject
+                WPAnalytics.track(.appSettingsImageQualityChanged, properties: properties)
```
Because [this tracking function](https://github.com/wordpress-mobile/WordPress-iOS/blob/3bc4737bd27414f5509e22a4caf83f601e3272d7/WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift#L1476) takes in `[AnyHashable: Any]` we should be able to do what was done in [`imageOptimizationChanged`](https://github.com/wordpress-mobile/WordPress-iOS/blob/3bc4737bd27414f5509e22a4caf83f601e3272d7/WordPress/Classes/Utility/Analytics/WPAnalyticsEvent.swift#L1476) after the suggestion is applied. `String` types will "just work" because they conform to `AnyHashable` so no need to cast. 👍

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/f29f4d2837d4a1d17b2d4f182f178680d8df74b6.

twstokes reviewed (commented):
Thanks for your work so far @fluiddot! I'm not entirely done and would like to focus on the table updates next, but I've left some comments for mostly minor things. 👍 

For the quality picker I wonder if we should add a short description of what this setting adjusts. We could loop @osullivanchris in to get thoughts. I also wondered about order, with the word "Low" being the highest row, and if we should reverse it. I understand the video resolution picker is similar so ultimately they both should match.

<img width="300px" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/2092798/4590ad33-d31d-4231-be47-a503c3be8bf5" />

fluiddot commented:
👋  Heads up that I'm temporarily marking this PR as `DO NOT MERGE`. The reason for this is we are still discussing the value to use as the default Image Quality setting (pcdRpT-4tj-p2).

osullivanchris commented:
> For the quality picker I wonder if we should add a short description of what this setting adjusts. We could loop @osullivanchris in to get thoughts. I also wondered about order, with the word "Low" being the highest row, and if we should reverse it. I understand the video resolution picker is similar so ultimately they both should match.

thanks for the ping @twstokes . Yeah I would put the order in reverse, and a description somewhere of what this does. We could start with a simple table row footer.

I am confused when I see the menu initially, I have options for both max file size, and image quality. Why would I set a max upload size, when this optimiser should be sorting stuff for me? As a user I don't quite understand the relationship between the two. Maybe the 'max size' could be a secondary option in the same details view as the image quality option.

Also usually I feel like max size would be related to file size rather than pixel size?

Happy to be corrected!

fluiddot committed:
Merge pull request #22019 from wordpress-mobile/task/use-multiplication-sign-to-indicate-dimensions

Use multiplication sign to indicate dimensions

fluiddot committed:
Use Leave On option as preferred in image optimization alert

Additionally, the casting has been removed from the event properties passed when tracking `app_settings_optimize_images_popup_tapped`.

fluiddot committed:
Remove unused function from `ImageQuality` enum

fluiddot committed:
Remove casting from event properties passed in `AppSettingsViewController`

fluiddot commented:
> thanks for the ping @twstokes . Yeah I would put the order in reverse, and a description somewhere of what this does. We could start with a simple table row footer.

I'd like to share that I set this order based on the current implementation on Android.

| App Settings | Max Image Upload Size | Image Quality |
|--------|--------|--------|
| ![app-settings](https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/e3eeaa83-5875-49e1-85a9-fa5602cef46e) | ![max-image-size](https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/157ddc3d-9afa-4461-a07f-e84cdf8839a4) | ![image-quality](https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/81790744-63ba-4e07-a4cf-6931d9ad3422) | 

Additionally, we already use this order in "Max Video Upload Size":
<img src=https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/44098d77-3829-4414-ac2e-fd47d5f5ce18 width=300>

However, we could inverse it for iOS if seems more suitable.

> I am confused when I see the menu initially, I have options for both max file size, and image quality. Why would I set a max upload size, when this optimiser should be sorting stuff for me? As a user I don't quite understand the relationship between the two. Maybe the 'max size' could be a secondary option in the same details view as the image quality option.

Good observation. I understand that the goal of presenting these two settings (i.e. Max Upload Size and Quality) is mainly to allow the user to customize the optimization. By default, the values will be the ones we consider optimal. Similar to my above comment, this implementation is based on what we currently have in the Android version.

Following your comments @osullivanchris, we could have an option below "Optimize Images" named "Optimization Settings" where we'd present these two settings. WDYT?

> Also usually I feel like max size would be related to file size rather than pixel size?

Good point, I shared a similar comment in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981#discussion_r1389765089. On Android this setting is displayed as "Maximum Image Size", I agree that size is usually more related to file size and not dimensions. Hence, we could rename the setting to something like "Max Image Upload Dimensions" or "Maximum Image Dimensions". WDYT?

**UPDATE:** For videos we also use the word `size`: "Max Video Upload Size". We might consider using `resolution` instead.

fluiddot committed:
Use unique reverse-DNS naming style in updated strings

fluiddot committed:
Rename constant `defaultMaxImageDimension`

fluiddot committed:
Rename `imageOptimizationSetting`

fluiddot committed:
Merge branch 'trunk' into add/image-optimization-setting

fluiddot committed:
Update release notes

fluiddot commented:
@twstokes @SiobhyB I've addressed all the feedback shared, hence the PR is ready for another review. I'm going to keep the `[Status] DO NOT MERGE` label until we decide the default optimization values for _Max Image Upload Size_ and _Image Quality_.

twstokes commented:
> hence the PR is ready for another review

On it! 👀

kean commented:
Heads up, @fluiddot: `MediaLibraryViewController` was removed and replaced with `SiteMediaViewController` in 23.7.

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -395,11 +449,23 @@ private extension AppSettingsViewController {
     func mediaTableSection() -> ImmuTableSection {
         let mediaHeader = NSLocalizedString("Media", comment: "Title label for the media settings section in the app settings")
 
+        let imageOptimizationValue = Bool(MediaSettings().imageOptimizationEnabled)
```
```suggestion
        let imageOptimizationValue = MediaSettings().imageOptimizationEnabled
```

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/b71b7f88ceea2abba57d36015fe8faf7e596dca0.

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -206,6 +201,39 @@ class AppSettingsViewController: UITableViewController {
+                          MediaSettings.ImageQuality.high,
+                          MediaSettings.ImageQuality.maximum]
+
+            let titles = values.map({ (settings: MediaSettings.ImageQuality) -> String in
```
Swift will infer the type being mapped over here (and returned), so if you wanted you could do:

`let titles = values.map { $0.description }`

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/b71b7f88ceea2abba57d36015fe8faf7e596dca0.

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -206,6 +201,39 @@ class AppSettingsViewController: UITableViewController {
+
+            let viewController = SettingsSelectionViewController(dictionary: settingsSelectionConfiguration)
+
+            viewController?.onItemSelected = { (quality: Any!) -> () in
```
```suggestion
            viewController?.onItemSelected = { quality in
```

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/b71b7f88ceea2abba57d36015fe8faf7e596dca0.

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -220,6 +248,32 @@ class AppSettingsViewController: UITableViewController {
+            WPAnalytics.track(.appSettingsOptimizeImagesChanged, properties: ["enabled": value])
+
+            // Show/hide image optimization settings
+            guard let self = self, let tableView = self.tableView else {
```
More Swift syntactic sugar...

```suggestion
            guard let self, let tableView else {
```

fluiddot replied:
That's interesting, I didn't know that Swift would automatically do this automatically, smart!

fluiddot replied:
Suggestion applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/b71b7f88ceea2abba57d36015fe8faf7e596dca0.

twstokes replied:
> I didn't know that Swift would automatically do this automatically, smart!

It's a [relatively new feature](https://github.com/apple/swift-evolution/blob/main/proposals/0345-if-let-shorthand.md)!

twstokes commented on WordPress/Classes/ViewRelated/Me/App Settings/AppSettingsViewController.swift:
```diff
@@ -220,6 +248,32 @@ class AppSettingsViewController: UITableViewController {
+            guard let self = self, let tableView = self.tableView else {
+                return
+            }
+            tableView.beginUpdates()
```
It looks like we can make this smoother if we set `handler.automaticallyReloadTableView = false` - maybe right after [it's set](https://github.com/wordpress-mobile/WordPress-iOS/blob/63d625037193b46fbaecc61701579d58b7f8076f/WordPress/Classes/ViewRelated/Me/App%20Settings/AppSettingsViewController.swift#L48). I think the call to `reloadData()` [in the setter](https://github.com/wordpress-mobile/WordPress-iOS/blob/63d625037193b46fbaecc61701579d58b7f8076f/WordPress/Classes/Utility/ImmuTable.swift#L260) is causing trouble.

After setting `automaticallyReloadTableView` to `false` as noted above, and doing the following:

```swift
            tableView.performBatchUpdates {
                self.reloadViewModel()
                let imageSizingIndex = IndexPath(row: 1, section: 0)
                let imageQualityIndex = IndexPath(row: 2, section: 0)
                let imageOptimizationSettingsIndexes = [imageSizingIndex, imageQualityIndex]
                tableView.reloadSections(IndexSet(integer: 0), with: .automatic)
            }
```

The animations worked as expected:

https://github.com/wordpress-mobile/WordPress-iOS/assets/2092798/f286bf50-64ef-4a3c-b943-6d9e69100c4e

⚠️ However, I don't know how disabling automatically reloading will affect existing functionality in the view. If we spot any issues we could just remove animations for now and address them later.

fluiddot replied:
> It looks like we can make this smoother if we set `handler.automaticallyReloadTableView = false` - maybe right after [it's set](https://github.com/wordpress-mobile/WordPress-iOS/blob/63d625037193b46fbaecc61701579d58b7f8076f/WordPress/Classes/ViewRelated/Me/App%20Settings/AppSettingsViewController.swift#L48). I think the call to `reloadData()` [in the setter](https://github.com/wordpress-mobile/WordPress-iOS/blob/63d625037193b46fbaecc61701579d58b7f8076f/WordPress/Classes/Utility/ImmuTable.swift#L260) is causing trouble.

I tried to set `automaticallyReloadTableView` to `false` after [it's set](https://github.com/wordpress-mobile/WordPress-iOS/blob/63d625037193b46fbaecc61701579d58b7f8076f/WordPress/Classes/ViewRelated/Me/App%20Settings/AppSettingsViewController.swift#L48) but leads to omit updating values in the table after changing them.

**Example:**

https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/e704cbb3-0b4f-4b5a-be4d-ac39b25dd9cf

_Note that _Image Quality_ and _Max Video Upload Size_ settings don't reflect the value selected._

The approach I found that works is to disable `automaticallyReloadTableView` before reloading the view model and set it to the original value after that. Changes applied in https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/132bfbcc2da4ec94a36a7caa4a6e285de06aa484.

https://github.com/wordpress-mobile/WordPress-iOS/assets/14905380/6f5735d5-4faa-43c1-ab0f-2ad8debf67c6

twstokes replied:
Good spot @fluiddot! Works for me. 👍

twstokes reviewed (approved):
Great work @fluiddot! The code LGTM and I left some optional comments. 🚀

fluiddot committed:
Apply code improvements in `AppSettingsViewController`

fluiddot committed:
Improve show/hide animation of image optimization settings

fluiddot committed:
Merge branch 'trunk' into add/image-optimization-setting

# Conflicts:
#	WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift

fluiddot commented:
> Heads up, @fluiddot: `MediaLibraryViewController` was removed and replaced with `SiteMediaViewController` in 23.7.

Thanks @kean for the heads up 🙇. I also noticed this when solving the file conflicts with `trunk` (https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/commits/119b3251e18367267a1ab5f3adb1591db38d3e40).

fluiddot commented:
@twstokes I've applied the suggestions from your recent feedback. The PR is approved so I'd consider it ready to merge, but let me know if you'd like to perform a quick review before. Thanks for the help 🙇 !

fluiddot commented on WordPress/Classes/Services/MediaSettings.swift:
```diff
@@ -110,13 +151,19 @@ class MediaSettings: NSObject {
     }
 
+    var imageQualityForUpload: ImageQuality {
+        return imageOptimizationEnabled ? imageQualitySetting : .high
```
When image optimization is disabled, the quality used is `High` not `Maximum`. This behavior matches the previous implementation where `preferredImageCompressionQuality` was used:

https://github.com/wordpress-mobile/WordPress-iOS/blob/e7a35f173d73ef4c0f7e7089e6d2676433b9a4d9/WordPress/Classes/Services/MediaImportService.swift#L22

https://github.com/wordpress-mobile/WordPress-iOS/pull/21981/files#diff-c162ac46983f01e3b35e008aa8b553cc0cb2b010a84926da308ee9b875e35638L381

twstokes commented:
> @twstokes I've applied the suggestions from your recent feedback. The PR is approved so I'd consider it ready to merge, but let me know if you'd like to perform a quick review before. Thanks for the help 🙇 !

Let's merge @fluiddot! 🚀
